database/database.rs

```diff
@@ -87,35 +88,39 @@ impl<D> Database<D> {
         &self.name
     }
 
-    pub(super) fn reserve_write_transaction(&self, timeout_millis: u64) {
-        // TODO: TransactionError
+    pub(super) fn reserve_write_transaction(&self, timeout_millis: u64) -> Result<(), TransactionError> {
+        // TODO: try_lock() instead of lock()? If it's TryLockError::Poisoned (and crash) or TryLockError::WouldBlock (try again later / timeout). Which timeout to use?
         let mut guard = self.schema_write_transaction_exclusivity.lock().unwrap();
         let (has_schema_transaction, running_write_transactions, ref mut notify_queue) = *guard;
         if has_schema_transaction || !notify_queue.is_empty() {
             let (sender, receiver) = sync_channel::<()>(0);
             notify_queue.push_back(TransactionReservationRequest::Write(sender));
             drop(guard);
-            receiver.recv().unwrap();
-            // TODO: turn into TransactionError on timeout
+            receiver
+                .recv_timeout(Duration::from_millis(timeout_millis))
+                .map_err(|source| TransactionError::Timeout { source })?;
```

Comment:
How does the other side of the queue handle the `SendError`s which will pop up if this timeout triggers?

Reply:
Good point, I didn't think there would be an error in this case. There was a 
`// TODO: if get disconnect error, skip!`

So I implemented it like
`let _skipped_sync_error = notifier.send(()).ok();`
Check the latest commits, please! I might have gotten the idea wrong.